File: src/main/java/pers/neige/neigeitems/action/node/impl/GradientParser.java
package pers.neige.neigeitems.action.node.impl;

import kotlin.text.StringsKt;
import lombok.NonNull;
import lombok.val;
import org.jetbrains.annotations.Nullable;
import pers.neige.neigeitems.action.ActionContext;
import pers.neige.neigeitems.action.node.NodeParser;
import pers.neige.neigeitems.config.ConfigReader;
import pers.neige.neigeitems.manager.BaseActionManager;
import pers.neige.neigeitems.utils.ColorUtils;
import pers.neige.neigeitems.utils.ListUtils;
import pers.neige.neigeitems.utils.NumberParser;
import pers.neige.neigeitems.utils.StringUtils;

public class GradientParser extends NodeParser {
    public GradientParser(@NonNull BaseActionManager manager) {
        super(manager);
    }

    @Override
    public @NonNull String getId() {
        return "gradient";
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull ConfigReader params
    ) {
        return handle(
            getParsedValue(context, params, "colorStart"),
            getParsedValue(context, params, "colorEnd"),
            getParsedValue(context, params, "step"),
            getParsedValue(context, params, "text")
        );
    }

    @Override
    public @Nullable String parse(
        @NonNull ActionContext context,
        @NonNull String params
    ) {
        val paramsList = StringUtils.split(params, '_', '\\', 4);
        return handle(
            paramsList.get(0),
            ListUtils.getOrNull(paramsList, 1),
            ListUtils.getOrNull(paramsList, 2),
            ListUtils.getOrNull(paramsList, 3)
        );
    }

    private @Nullable String handle(
        @Nullable String colorStartString,
        @Nullable String colorEndString,
        @Nullable String stepString,
        @Nullable String text
    ) {
        if (colorStartString == null) {
            warning("未指定起始颜色");
            return null;
        }
        if (colorEndString == null) {
            warning("未指定终止颜色");
            return null;
        }
        if (text == null) {
            warning("未指定待处理文本");
            return null;
        }
        Integer colorStart = parse(colorStartString, 0, (it) -> StringsKt.toIntOrNull(it, 16), "{0} 并非数字, 无法用作起始颜色");
        if (colorStart == null) return null;
        colorStart = Math.min(Math.max(colorStart, 0), 0xFFFFFF);
        Integer colorEnd = parse(colorEndString, 0, (it) -> StringsKt.toIntOrNull(it, 16), "{0} 并非数字, 无法用作终止颜色");
        if (colorEnd == null) return null;
        colorEnd = Math.min(Math.max(colorEnd, 0), 0xFFFFFF);
        val step = parse(stepString, 1, NumberParser::parseInteger, "{0} 并非数字, 无法用作步长");
        if (step == null) return null;

        if (text.length() <= step) {

            return ColorUtils.toHexColorPrefix(colorStart) + text;
        }

        val chars = text.toCharArray();
        val result = new StringBuilder();

        int redCurrent = ColorUtils.getRed(colorStart);
        int greenCurrent = ColorUtils.getGreen(colorStart);
        int blueCurrent = ColorUtils.getBlue(colorStart);

        if (step == 1) {
            val redStep = (ColorUtils.getRed(colorEnd) - redCurrent) / (chars.length - 1);
            val greenStep = (ColorUtils.getGreen(colorEnd) - greenCurrent) / (chars.length - 1);
            val blueStep = (ColorUtils.getBlue(colorEnd) - blueCurrent) / (chars.length - 1);

            for (val c : chars) {
                result.append(ColorUtils.toHexColorPrefix(redCurrent, greenCurrent, blueCurrent));
                result.append(c);
                redCurrent += redStep;
                greenCurrent += greenStep;
                blueCurrent += blueStep;
            }
        } else {
            val redStep = (ColorUtils.getRed(colorEnd) - redCurrent) * step / (chars.length - 1);
            val greenStep = (ColorUtils.getGreen(colorEnd) - greenCurrent) * step / (chars.length - 1);
            val blueStep = (ColorUtils.getBlue(colorEnd) - blueCurrent) * step / (chars.length - 1);

            int current = 1;

            for (val c : chars) {
                if (current == 1) {
                    result.append(ColorUtils.toHexColorPrefix(redCurrent, greenCurrent, blueCurrent));
                    redCurrent += redStep;
                    greenCurrent += greenStep;
                    blueCurrent += blueStep;
                }
                result.append(c);
                if (current == step) {
                    current = 1;
                } else {
                    current++;
                }
            }
        }
        return result.toString();
    }
}

File: src/main/kotlin/pers/neige/neigeitems/section/impl/GradientParser.kt
package pers.neige.neigeitems.section.impl

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.ConfigurationSection
import pers.neige.neigeitems.section.SectionParser
import pers.neige.neigeitems.utils.ColorUtils
import pers.neige.neigeitems.utils.SectionUtils.parseSection
import pers.neige.neigeitems.utils.StringUtils.joinToString

/**
 * 渐变色节点解析器
 */
object GradientParser : SectionParser() {
    override val id: String = "gradient"

    override fun onRequest(
        data: ConfigurationSection,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String? {
        return handler(
            cache,
            player,
            sections,
            true,
            data.getString("colorStart"),
            data.getString("colorEnd"),
            data.getString("step"),
            data.getString("text")
        )
    }

    override fun onRequest(
        args: List<String>,
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?
    ): String {
        return handler(
            cache,
            player,
            sections,
            false,
            args.getOrNull(0),
            args.getOrNull(1),
            args.getOrNull(2),
            args.joinToString("_", 3)
        ) ?: "<$id::${args.joinToString("_")}>"
    }

    /**
     * @param cache 解析值缓存
     * @param player 待解析玩家
     * @param sections 节点池
     * @param parse 是否对参数进行节点解析
     * @param colorStartString 起始颜色
     * @param colorEndString 结尾颜色
     * @param stepString 每几个字符变一次色
     * @param rawText 文本内容
     * @return 解析值
     */
    private fun handler(
        cache: MutableMap<String, String>?,
        player: OfflinePlayer?,
        sections: ConfigurationSection?,
        parse: Boolean,
        colorStartString: String?,
        colorEndString: String?,
        stepString: String?,
        rawText: String?
    ): String? {
        if (colorStartString == null || colorEndString == null || rawText == null) return null
        val colorStart =
            (colorStartString.parseSection(parse, cache, player, sections).toIntOrNull(16) ?: 0)
                .coerceAtLeast(0)
                .coerceAtMost(0xFFFFFF)
        val colorEnd =
            (colorEndString.parseSection(parse, cache, player, sections).toIntOrNull(16) ?: 0)
                .coerceAtLeast(0)
                .coerceAtMost(0xFFFFFF)

        val step = (stepString?.parseSection(parse, cache, player, sections)?.toIntOrNull() ?: 1)
            .coerceAtLeast(1)

        val text = rawText.parseSection(parse, cache, player, sections)

        if (text.length <= step) {

            return ColorUtils.toHexColorPrefix(colorStart) + text
        }

        val chars = text.toCharArray()
        val result = StringBuilder()

        var redCurrent = ColorUtils.getRed(colorStart)
        var greenCurrent = ColorUtils.getGreen(colorStart)
        var blueCurrent = ColorUtils.getBlue(colorStart)

        if (step == 1) {
            val redStep = (ColorUtils.getRed(colorEnd) - redCurrent) / chars.lastIndex
            val greenStep = (ColorUtils.getGreen(colorEnd) - greenCurrent) / chars.lastIndex
            val blueStep = (ColorUtils.getBlue(colorEnd) - blueCurrent) / chars.lastIndex

            for (char in chars) {
                result.append(ColorUtils.toHexColorPrefix(redCurrent, greenCurrent, blueCurrent))
                result.append(char)
                redCurrent += redStep
                greenCurrent += greenStep
                blueCurrent += blueStep
            }
        } else {
            val redStep = (ColorUtils.getRed(colorEnd) - redCurrent) * step / chars.lastIndex
            val greenStep = (ColorUtils.getGreen(colorEnd) - greenCurrent) * step / chars.lastIndex
            val blueStep = (ColorUtils.getBlue(colorEnd) - blueCurrent) * step / chars.lastIndex

            var current = 1

            for (char in chars) {
                if (current == 1) {
                    result.append(ColorUtils.toHexColorPrefix(redCurrent, greenCurrent, blueCurrent))
                    redCurrent += redStep
                    greenCurrent += greenStep
                    blueCurrent += blueStep
                }
                result.append(char)
                if (current == step) {
                    current = 1
                } else {
                    current++
                }
            }
        }
        return result.toString()
    }
}